archunit/src/main/java/com/tngtech/archunit/library/freeze/ViolationLineMatcherFactory.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2019 TNG Technology Consulting GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.tngtech.archunit.library.freeze;
+
+import java.util.Iterator;
+
+import com.google.common.base.Splitter;
+import com.tngtech.archunit.ArchConfiguration;
+import com.tngtech.archunit.core.MayResolveTypesViaReflection;
+
+import static com.tngtech.archunit.base.ReflectionUtils.newInstanceOf;
+import static java.lang.Character.isDigit;
+
+class ViolationLineMatcherFactory {
+    private static final String FREEZE_LINE_MATCHER_PROPERTY = "freeze.lineMatcher";
+    private static final FuzzyLineNumberMatcher DEFAULT_MATCHER = new FuzzyLineNumberMatcher();
+
+    static ViolationLineMatcher create() {
+        return ArchConfiguration.get().containsProperty(FREEZE_LINE_MATCHER_PROPERTY)
+                ? createInstance(ArchConfiguration.get().getProperty(FREEZE_LINE_MATCHER_PROPERTY))
+                : DEFAULT_MATCHER;
+    }
+
+    @MayResolveTypesViaReflection(reason = "This is not part of the import process")
+    private static ViolationLineMatcher createInstance(String lineMatcherClassName) {
+        try {
+            return (ViolationLineMatcher) newInstanceOf(Class.forName(lineMatcherClassName));
+        } catch (Exception e) {
+            String message = String.format("Could not instantiate %s of configured type '%s=%s'",
+                    ViolationLineMatcher.class.getSimpleName(), FREEZE_LINE_MATCHER_PROPERTY, lineMatcherClassName);
+            throw new StoreInitializationFailedException(message, e);
+        }
+    }
+
+    private static class FuzzyLineNumberMatcher implements ViolationLineMatcher {
+        @Override
+        public boolean matches(String lineFromFirstViolation, String lineFromSecondViolation) {
+            return ignoreLineNumber(lineFromFirstViolation).equals(ignoreLineNumber(lineFromSecondViolation));
+        }
+
+        // Would be nicer to use a regex here, but unfortunately that would have a massive performance impact.
+        // So we do some low level character processing instead.
+        private String ignoreLineNumber(String violation) {
```

Comment:
A stringly typed violation API cannot guarantee to ignore only line numbers.
You're just matching the pattern of `SourceCodeLocation` strings being `format("(%s:%d)", sourceFileName, lineNumber)`. (But no doubt that this is reasonable to do!)

I was wondering whether you could gain some performance by skipping your algorithm unless `(violation.contains("(") && violation.contains(")")`? Maybe not, as probably most of ArchUnit's conditions create events whose description lines contain such a `SourceCodeLocation`, so only custom conditions could at all benefit from that.

But what about an early exit in `matches` when `lineFromFirstViolation.equals(lineFromSecondViolation)` already?
I could imagine that some violations might indeed not change at all when `FreezingArchRule` is applied to a legacy project with tons of violations.